File: go/pkg/dashscope/event.go
package dashscope

// Event types for realtime communication.
const (
	// Client events
	EventTypeSessionUpdate       = "session.update"
	EventTypeInputAudioAppend    = "input_audio_buffer.append"
	EventTypeInputAudioCommit    = "input_audio_buffer.commit"
	EventTypeInputAudioClear     = "input_audio_buffer.clear"
	EventTypeResponseCreate      = "response.create"
	EventTypeResponseCancel      = "response.cancel"
	EventTypeTranscriptionUpdate = "transcription.update"

	// Server events
	EventTypeSessionCreated          = "session.created"
	EventTypeSessionUpdated          = "session.updated"
	EventTypeInputAudioCommitted     = "input_audio_buffer.committed"
	EventTypeInputAudioCleared       = "input_audio_buffer.cleared"
	EventTypeInputSpeechStarted      = "input_audio_buffer.speech_started"
	EventTypeInputSpeechStopped      = "input_audio_buffer.speech_stopped"
	EventTypeResponseCreated         = "response.created"
	EventTypeResponseDone            = "response.done"
	EventTypeResponseOutputAdded     = "response.output_item.added"
	EventTypeResponseOutputDone      = "response.output_item.done"
	EventTypeResponseContentAdded    = "response.content_part.added"
	EventTypeResponseContentDone     = "response.content_part.done"
	EventTypeResponseTextDelta       = "response.text.delta"
	EventTypeResponseTextDone        = "response.text.done"
	EventTypeResponseAudioDelta      = "response.audio.delta"
	EventTypeResponseAudioDone       = "response.audio.done"
	EventTypeResponseTranscriptDelta = "response.audio_transcript.delta"
	EventTypeResponseTranscriptDone  = "response.audio_transcript.done"
	EventTypeError                   = "error"

	// DashScope-specific: "choices" format response (different from OpenAI Realtime)
	EventTypeChoicesResponse = "choices"
)

// RealtimeEvent represents an event in the realtime session.
type RealtimeEvent struct {
	// Type is the event type.
	// For DashScope "choices" format responses, this will be EventTypeChoicesResponse.
	Type string `json:"type"`

	// EventID is the unique identifier for this event.
	EventID string `json:"event_id,omitempty"`

	// Session contains session information (for session.* events).
	Session *SessionInfo `json:"session,omitempty"`

	// Response contains response information (for response.* events).
	Response *ResponseInfo `json:"response,omitempty"`

	// ResponseID is the response identifier (for response.created).
	ResponseID string `json:"response_id,omitempty"`

	// Delta contains incremental text content (for *.delta events and choices responses).
	Delta string `json:"delta,omitempty"`

	// Audio contains decoded audio data (for response.audio.delta and choices responses).
	Audio []byte `json:"-"`

	// AudioBase64 is the raw base64 audio from JSON.
	AudioBase64 string `json:"audio,omitempty"`

	// Transcript contains transcript text (for transcript completion events).
	Transcript string `json:"transcript,omitempty"`

	// FinishReason indicates why the response ended (for choices responses).
	// Values: "stop", "length", etc. Empty string means still generating.
	FinishReason string `json:"finish_reason,omitempty"`

	// ItemID is the item identifier (for item events).
	ItemID string `json:"item_id,omitempty"`

	// OutputIndex is the output index (for content events).
	OutputIndex int `json:"output_index,omitempty"`

	// ContentIndex is the content index (for content events).
	ContentIndex int `json:"content_index,omitempty"`

	// Error contains error information (for error events).
	Error *EventError `json:"error,omitempty"`

	// Usage contains usage statistics (for response.done).
	Usage *UsageStats `json:"usage,omitempty"`
}

// ResponseInfo contains response state information.
type ResponseInfo struct {
	ID           string        `json:"id,omitempty"`
	Status       string        `json:"status,omitempty"`
	StatusDetail *StatusDetail `json:"status_detail,omitempty"`
	Output       []OutputItem  `json:"output,omitempty"`
	Usage        *UsageStats   `json:"usage,omitempty"`
}

// StatusDetail contains detailed status information.
type StatusDetail struct {
	Type   string `json:"type,omitempty"`
	Reason string `json:"reason,omitempty"`
	Error  *Error `json:"error,omitempty"`
}

// OutputItem represents an output item in a response.
type OutputItem struct {
	ID      string        `json:"id,omitempty"`
	Type    string        `json:"type,omitempty"`
	Role    string        `json:"role,omitempty"`
	Status  string        `json:"status,omitempty"`
	Content []ContentPart `json:"content,omitempty"`
}

// ContentPart represents a part of content.
type ContentPart struct {
	Type       string `json:"type,omitempty"`
	Text       string `json:"text,omitempty"`
	Audio      string `json:"audio,omitempty"`
	Transcript string `json:"transcript,omitempty"`
}

// EventError contains error information from error events.
type EventError struct {
	Type    string `json:"type,omitempty"`
	Code    string `json:"code,omitempty"`
	Message string `json:"message,omitempty"`
	Param   string `json:"param,omitempty"`
}

File: rust/dashscope/src/event.rs
//! Event types for realtime communication.

use serde::{Deserialize, Serialize};

use crate::types::{SessionInfo, UsageStats};

/// Client events
pub const EVENT_TYPE_SESSION_UPDATE: &str = "session.update";
pub const EVENT_TYPE_INPUT_AUDIO_APPEND: &str = "input_audio_buffer.append";
pub const EVENT_TYPE_INPUT_AUDIO_COMMIT: &str = "input_audio_buffer.commit";
pub const EVENT_TYPE_INPUT_AUDIO_CLEAR: &str = "input_audio_buffer.clear";
pub const EVENT_TYPE_INPUT_IMAGE_APPEND: &str = "input_image_buffer.append";
pub const EVENT_TYPE_RESPONSE_CREATE: &str = "response.create";
pub const EVENT_TYPE_RESPONSE_CANCEL: &str = "response.cancel";
pub const EVENT_TYPE_SESSION_FINISH: &str = "session.finish";

/// Server events
pub const EVENT_TYPE_SESSION_CREATED: &str = "session.created";
pub const EVENT_TYPE_SESSION_UPDATED: &str = "session.updated";
pub const EVENT_TYPE_INPUT_AUDIO_COMMITTED: &str = "input_audio_buffer.committed";
pub const EVENT_TYPE_INPUT_AUDIO_CLEARED: &str = "input_audio_buffer.cleared";
pub const EVENT_TYPE_INPUT_SPEECH_STARTED: &str = "input_audio_buffer.speech_started";
pub const EVENT_TYPE_INPUT_SPEECH_STOPPED: &str = "input_audio_buffer.speech_stopped";
pub const EVENT_TYPE_RESPONSE_CREATED: &str = "response.created";
pub const EVENT_TYPE_RESPONSE_DONE: &str = "response.done";
pub const EVENT_TYPE_RESPONSE_OUTPUT_ADDED: &str = "response.output_item.added";
pub const EVENT_TYPE_RESPONSE_OUTPUT_DONE: &str = "response.output_item.done";
pub const EVENT_TYPE_RESPONSE_CONTENT_ADDED: &str = "response.content_part.added";
pub const EVENT_TYPE_RESPONSE_CONTENT_DONE: &str = "response.content_part.done";
pub const EVENT_TYPE_RESPONSE_TEXT_DELTA: &str = "response.text.delta";
pub const EVENT_TYPE_RESPONSE_TEXT_DONE: &str = "response.text.done";
pub const EVENT_TYPE_RESPONSE_AUDIO_DELTA: &str = "response.audio.delta";
pub const EVENT_TYPE_RESPONSE_AUDIO_DONE: &str = "response.audio.done";
pub const EVENT_TYPE_RESPONSE_TRANSCRIPT_DELTA: &str = "response.audio_transcript.delta";
pub const EVENT_TYPE_RESPONSE_TRANSCRIPT_DONE: &str = "response.audio_transcript.done";
pub const EVENT_TYPE_INPUT_AUDIO_TRANSCRIPTION_COMPLETED: &str = "conversation.item.input_audio_transcription.completed";
pub const EVENT_TYPE_ERROR: &str = "error";

/// DashScope-specific: "choices" format response
pub const EVENT_TYPE_CHOICES_RESPONSE: &str = "choices";

/// Realtime event from the server.
#[derive(Debug, Clone, Default)]
pub struct RealtimeEvent {
    /// Event type.
    pub event_type: String,

    /// Unique event identifier.
    pub event_id: Option<String>,

    /// Session information (for session.* events).
    pub session: Option<SessionInfo>,

    /// Response information (for response.* events).
    pub response: Option<ResponseInfo>,

    /// Response identifier (for response.created).
    pub response_id: Option<String>,

    /// Incremental text content (for *.delta events).
    pub delta: Option<String>,

    /// Decoded audio data (for response.audio.delta and choices responses).
    pub audio: Option<Vec<u8>>,

    /// Raw base64 audio from JSON.
    pub audio_base64: Option<String>,

    /// Transcript text (for transcript completion events).
    pub transcript: Option<String>,

    /// Finish reason (for choices responses).
    pub finish_reason: Option<String>,

    /// Item identifier (for item events).
    pub item_id: Option<String>,

    /// Output index (for content events).
    pub output_index: Option<i32>,

    /// Content index (for content events).
    pub content_index: Option<i32>,

    /// Error information (for error events).
    pub error: Option<EventError>,

    /// Usage statistics (for response.done).
    pub usage: Option<UsageStats>,
}

/// Response information.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ResponseInfo {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status_detail: Option<StatusDetail>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub output: Option<Vec<OutputItem>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub usage: Option<UsageStats>,
}

/// Status detail information.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct StatusDetail {
    #[serde(rename = "type", skip_serializing_if = "Option::is_none")]
    pub detail_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<EventError>,
}

/// Output item in a response.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct OutputItem {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub id: Option<String>,
    #[serde(rename = "type", skip_serializing_if = "Option::is_none")]
    pub item_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub role: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub status: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub content: Option<Vec<ContentPart>>,
}

/// Content part.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct ContentPart {
    #[serde(rename = "type", skip_serializing_if = "Option::is_none")]
    pub content_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub text: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub audio: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub transcript: Option<String>,
}

/// Error information from error events.
#[derive(Debug, Clone, Default, Serialize, Deserialize)]
pub struct EventError {
    #[serde(rename = "type", skip_serializing_if = "Option::is_none")]
    pub error_type: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub message: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub param: Option<String>,
}

impl RealtimeEvent {
    /// Returns true if this is a session.created event.
    pub fn is_session_created(&self) -> bool {
        self.event_type == EVENT_TYPE_SESSION_CREATED
    }

    /// Returns true if this is a session.updated event.
    pub fn is_session_updated(&self) -> bool {
        self.event_type == EVENT_TYPE_SESSION_UPDATED
    }

    /// Returns true if this is a response.done event.
    pub fn is_response_done(&self) -> bool {
        self.event_type == EVENT_TYPE_RESPONSE_DONE
    }

    /// Returns true if this is an error event.
    pub fn is_error(&self) -> bool {
        self.event_type == EVENT_TYPE_ERROR
    }

    /// Returns true if this is a choices response (DashScope-specific).
    pub fn is_choices_response(&self) -> bool {
        self.event_type == EVENT_TYPE_CHOICES_RESPONSE
    }

    /// Returns true if this event contains audio data.
    pub fn has_audio(&self) -> bool {
        self.audio.is_some() || self.audio_base64.is_some()
    }

    /// Returns true if this event contains text delta.
    pub fn has_text_delta(&self) -> bool {
        self.delta.is_some()
    }
}
